### ISSUE TITLE
[bug]: Parse error of $args in dialparties.agi - causing call confirm to break.

### ISSUE THREAD

**Issue by ravnx** [FreePBX/issue-tracker#433]
### FreePBX Version

FreePBX 17

### Issue Description

Having an issue where when call confirm is on and a cell phone is in a ring group, or the call is call forwarded, and call forwarding is allowed, the dialparties.agi has a parse error.

Example:

> [2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Added extension { to extension map
> [2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Added extension 7135689715#} to extension map
> [2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension { cf is disabled
> [2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension 7135689715#} cf is disabled
> [2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension { do not disturb is disabled
> [2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension 7135689715#} do not disturb is disabled
> [2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: EXTENSION_STATE: 4 (UNKNOWN)

The line:  $extarray = preg_split( '/-/', $arg ); ends up splitting wrong because $arg will look like "{-7135689715#}"

Around line 181 of dialparties.agi I just added a bandaid to fix this issue, but not sure where the root cause is:
        # clean the { } characters out of $arg # mpalmer
        $arg = str_replace(array('{','}'), '', $arg);

### Operating Environment

Debian 6.1.0-22-amd64
Asterisk 21.4.3

```
+---------------------+-----------+-----------------------------------+-------------+-----------+
| Module              | Version   | Status                            | License     | Signature |
+---------------------+-----------+-----------------------------------+-------------+-----------+
| accountcodepreserve | 17.0.0.1  | Enabled                           | GPLv2       | Sangoma   |
| adv_recovery        |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| allowlist           | 17.0.1.1  | Enabled                           | GPLv3+      | Sangoma   |
| amd                 | 17.0.1    | Enabled                           | GPLv3+      | Sangoma   |
| announcement        | 17.0.2.1  | Enabled                           | GPLv3+      | Sangoma   |
| api                 | 17.0.1.6  | Enabled                           | AGPLv3+     | Sangoma   |
| areminder           |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| arimanager          | 17.0.1.1  | Enabled                           | GPLv3+      | Sangoma   |
| asterisk-cli        | 17.0.2    | Enabled                           | GPLv3+      | Sangoma   |
| asteriskinfo        | 17.0.2    | Enabled                           | GPLv3+      | Sangoma   |
| backup              | 17.0.5.60 | Enabled                           | GPLv3+      | Sangoma   |
| blacklist           | 17.0.1.2  | Enabled                           | GPLv3+      | Sangoma   |
| broadcast           |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| builtin             |           | Enabled                           |             | Unsigned  |
| bulkhandler         | 17.0.5    | Enabled                           | GPLv3+      | Sangoma   |
| calendar            | 17.0.4.19 | Enabled                           | GPLv3+      | Sangoma   |
| callaccounting      | 17.0.5    | Enabled                           | Commercial+ | Sangoma   |
| callback            | 17.0.2.1  | Enabled                           | GPLv3+      | Sangoma   |
| callerid            |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| callforward         | 17.0.1.6  | Enabled                           | AGPLv3+     | Sangoma   |
| calllimit           |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| callrecording       | 17.0.3.7  | Enabled                           | AGPLv3+     | Sangoma   |
| callwaiting         | 17.0.3.4  | Enabled                           | GPLv3+      | Sangoma   |
| cdr                 | 17.0.4.20 | Enabled                           | GPLv3+      | Sangoma   |
| cdrpro              |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| cel                 | 17.0.2.8  | Enabled                           | GPLv3+      | Sangoma   |
| certman             | 17.0.3.13 | Enabled                           | AGPLv3+     | Sangoma   |
| cidlookup           | 17.0.1.1  | Enabled                           | GPLv3+      | Sangoma   |
| conferences         | 17.0.3.2  | Enabled                           | GPLv3+      | Sangoma   |
| conferencespro      |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| configedit          | 17.0.1.4  | Enabled                           | AGPLv3+     | Sangoma   |
| contactmanager      | 17.0.5.11 | Enabled                           | GPLv3+      | Sangoma   |
| core                | 17.0.16   | Enabled                           | GPLv3+      | Sangoma   |
| cos                 | 17.0.1.1  | Enabled                           | Commercial  | Sangoma   |
| customappsreg       | 17.0.1    | Enabled                           | GPLv3+      | Sangoma   |
| customcontexts      | 17.0.1.3  | Enabled                           | GPLv2+      | Sangoma   |
| dashboard           | 17.0.4.5  | Enabled                           | AGPLv3+     | Sangoma   |
| daynight            | 17.0.1.1  | Enabled                           | GPLv3+      | Sangoma   |
| dictate             |           | Not Installed (Locally available) | GPLv3+      | Sangoma   |
| directory           | 17.0.1.1  | Enabled                           | GPLv3+      | Sangoma   |
| disa                | 17.0.5    | Enabled                           | AGPLv3+     | Sangoma   |
| donotdisturb        | 17.0.2.3  | Enabled                           | GPLv3+      | Sangoma   |
| dynroute            | 17.0.3.2  | Enabled                           | GPLv3+      | Sangoma   |
| endpoint            | 17.0.1.81 | Enabled                           | Commercial  | Sangoma   |
| extensionroutes     | 17.0.1    | Enabled                           | Commercial  | Sangoma   |
| extensionsettings   | 17.0.1    | Enabled                           | GPLv3+      | Sangoma   |
| fax                 | 17.0.3.4  | Enabled                           | GPLv3+      | Sangoma   |
| faxpro              |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| featurecodeadmin    | 17.0.2    | Enabled                           | GPLv3+      | Sangoma   |
| filestore           | 17.0.2.23 | Enabled                           | AGPLv3      | Sangoma   |
| findmefollow        | 17.0.4.10 | Enabled                           | GPLv3+      | Sangoma   |
| firewall            | 17.0.1.26 | Enabled                           | AGPLv3+     | Sangoma   |
| framework           | 17.0.19.9 | Enabled                           | GPLv2+      | Sangoma   |
| hotelwakeup         | 17.0.1.6  | Enabled                           | GPLv2       | Sangoma   |
| iaxsettings         | 17.0.1    | Enabled                           | AGPLv3      | Sangoma   |
| infoservices        | 17.0.1.1  | Enabled                           | GPLv2+      | Sangoma   |
| ivr                 | 17.0.8    | Enabled                           | GPLv3+      | Sangoma   |
| languages           | 17.0.1    | Enabled                           | GPLv3+      | Sangoma   |
| logfiles            | 17.0.3.3  | Enabled                           | GPLv3+      | Sangoma   |
| manager             | 17.0.6    | Enabled                           | GPLv2+      | Sangoma   |
| miscapps            | 17.0.3    | Enabled                           | GPLv3+      | Sangoma   |
| miscdests           | 17.0.1.1  | Enabled                           | GPLv3+      | Sangoma   |
| missedcall          | 17.0.1.1  | Enabled                           | GPLv3+      | Sangoma   |
| music               | 17.0.5    | Enabled                           | GPLv3+      | Sangoma   |
| outcnam             | 17.0.1    | Enabled                           | GPLv3+      | Sangoma   |
| outroutemsg         | 17.0.1    | Enabled                           | GPLv3+      | Sangoma   |
| paging              | 17.0.3    | Enabled                           | GPLv3+      | Sangoma   |
| pagingpro           |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| parking             | 17.0.2.5  | Enabled                           | GPLv3+      | Sangoma   |
| parkpro             |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| pbxmfa              | 17.0.2.1  | Enabled                           | Commercial+ | Sangoma   |
| phpinfo             | 17.0.1    | Enabled                           | GPLv2+      | Sangoma   |
| pinsets             | 17.0.3.2  | Enabled                           | GPLv3+      | Sangoma   |
| pinsetspro          |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| pm2                 | 17.0.3.2  | Enabled                           | AGPLv3+     | Sangoma   |
| pms                 | 17.0.2.18 | Enabled                           | Commercial  | Sangoma   |
| presencestate       | 17.0.2.4  | Enabled                           | GPLv3+      | Sangoma   |
| printextensions     | 17.0.1.2  | Enabled                           | GPLv3+      | Sangoma   |
| queueprio           | 17.0.1.4  | Enabled                           | GPLv3+      | Sangoma   |
| queues              | 17.0.1.11 | Enabled                           | GPLv2+      | Sangoma   |
| queuestats          |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| qxact_reports       |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| recording_report    |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| recordings          | 17.0.2.2  | Enabled                           | GPLv3+      | Sangoma   |
| restapps            | 17.0.1.21 | Enabled                           | Commercial  | Sangoma   |
| ringgroups          | 17.0.2.4  | Enabled                           | GPLv3+      | Sangoma   |
| sangomaconnect      | 17.0.1.35 | Enabled                           | Commercial  | Sangoma   |
| sangomacrm          |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| sangomartapi        | 17.0.1.28 | Enabled                           | Commercial  | Sangoma   |
| setcid              | 17.0.1.2  | Enabled                           | GPLv3+      | Sangoma   |
| sipsettings         | 17.0.6.9  | Enabled                           | AGPLv3+     | Sangoma   |
| sipstation          |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| sms                 | 17.0.1.15 | Enabled                           | Commercial  | Sangoma   |
| smsplus             | 17.0.3    | Enabled                           | Commercial  | Sangoma   |
| soundlang           | 17.0.4.1  | Enabled                           | GPLv3+      | Sangoma   |
| superfecta          | 17.0.2.1  | Enabled                           | GPLv2+      | Sangoma   |
| sysadmin            | 17.0.1.95 | Enabled                           | Commercial  | Sangoma   |
| timeconditions      | 17.0.1.18 | Enabled                           | GPLv3+      | Sangoma   |
| tts                 | 17.0.1.1  | Enabled                           | GPLv3+      | Sangoma   |
| ttsengines          | 17.0.1    | Enabled                           | AGPLv3      | Sangoma   |
| ucp                 | 17.0.4.21 | Enabled                           | AGPLv3+     | Sangoma   |
| userman             | 17.0.6.27 | Enabled                           | AGPLv3+     | Sangoma   |
| vmblast             |           | Not Installed (Locally available) | GPLv3+      | Sangoma   |
| vmnotify            |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| voicemail           | 17.0.5.22 | Enabled                           | GPLv3+      | Sangoma   |
| voicemail_report    |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| voipinnovations     |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| vqplus              |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| weakpasswords       | 17.0.1    | Enabled                           | GPLv3+      | Sangoma   |
| webcallback         |           | Not Installed (Locally available) | Commercial  | Sangoma   |
| webrtc              | 17.0.2.2  | Enabled                           | GPLv3+      | Sangoma   |
+---------------------+-----------+-----------------------------------+-------------+-----------+
```

### Relevant log output

```shell
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] pbx.c: Executing [s@macro-dial:5] ExecIf("Local/RG-930*-7135689715#@from-internal-000000fa;2", "0?Set(CHANNEL(musicclass)=)") in new stack
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] pbx.c: Executing [s@macro-dial:6] AGI("Local/RG-930*-7135689715#@from-internal-000000fa;2", "agi://127.0.0.1/dialparties.agi") in new stack
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Starting New Dialparties.agi
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Caller ID name is '2816507573' number is '2816507573'
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: CW Ignore is: 
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: CF Ignore is: 
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: CW IN_USE/BUSY is: 1
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Ringgroup confirm is  : 1
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Methodology of ring is  'none'
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension array: Array
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Added extension { to extension map
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Added extension 7135689715#} to extension map
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension { cf is disabled
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension 7135689715#} cf is disabled
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension { do not disturb is disabled
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension 7135689715#} do not disturb is disabled
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: EXTENSION_STATE: 4 (UNKNOWN)
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension { has ExtensionState: 4
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Checking CW and CFB status for extension {
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension { is not available to be called
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: EXTENSION_STATE: 4 (UNKNOWN)
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension 7135689715#} has ExtensionState: 4
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Checking CW and CFB status for extension 7135689715#}
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Extension 7135689715#} is not available to be called
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Filtered ARG3: 
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: Setting default NOANSWER DIALSTATUS since no extensions available
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: RVOL_MODE '' 
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: RVOL is: 
[2024-09-10 18:04:44] VERBOSE[1697609][C-0000019f] res_agi.c: agi://127.0.0.1/dialparties.agi: RVOLPARENT is: 
[2024-09-10 18:04:45] VERBOSE[1697609][C-0000019f] res_agi.c: <Local/RG-930*-7135689715#@from-internal-000000fa;2>AGI Script agi://127.0.0.1/dialparties.agi completed, returning 0
```

**Comment by KSRyanMerritt** [comment]
I am also having the exact same problem

root@pbx:/etc/apache2# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"

root@pbx:/etc/apache2# asterisk -V
Asterisk 18.24.1

    -- Executing [s@macro-dial:7] AGI("Local/RG-306*-479883994#@from-internal-00000001;2", "dialparties.agi") in new stack
    -- Launched AGI Script /var/lib/asterisk/agi-bin/dialparties.agi
    -- <Local/RG-306*-479883994#@from-internal-00000001;2>AGI Script dialparties.agi completed, returning 0
    -- Executing [s@macro-dial:8] GotoIf("Local/RG-306*-479883994#@from-internal-00000001;2", "0?") in new stack
    -- Executing [s@macro-dial:9] NoOp("Local/RG-306*-479883994#@from-internal-00000001;2", "Returned from dialparties with no extensions to call and DIALSTATUS: NOANSWER") in new stack
    -- Executing [s@macro-dial:10] Return("Local/RG-306*-479883994#@from-internal-00000001;2", "") in new stack
    -- Auto fallthrough, channel 'Local/RG-306*-479883994#@from-internal-00000001;2' status is 'NOANSWER'

**Comment by ravnx** [comment]
I've got a ticket open with Sangoma, but they're still bouncing around with requesting PCAPs and telling me my ring group is setup wrong. They're supposed to be escalating to dev team as well, there were some updates to core over the weekend pushed to public, so I will test again after hours removing my fix and letting dialparties.agi overwrite to see if its been fixed.

**Comment by kapilgupta01** [comment]
Hi All

just an update that we are working on this issue.

**Comment by jissphilip** [comment]
ringgroups version 17.0.2.5

**Comment by KSRyanMerritt** [comment]
I have tested and confirmed this is working now.  Thank you so much for your quick turnaround and hard work on this!

**Comment by KSRyanMerritt** [comment]
So it turns out I misspoke.  While it is playing the audio to the remote caller confirms the call by pressing 1.  When the remote caller presses 1 to take the call it sends them to voice mail.  It acts like you didn't press 1.

[2024-09-17 14:36:02] ERROR[1738389][C-0000015f]: app_stack.c:401 return_exec: Return without Gosub: stack is empty
  == Spawn extension (macro-confirm, 1, 10) exited non-zero on 'Local/479883994@from-internal-000000b3;1' in macro 'confirm'
    -- Local/479883994@from-internal-000000b3;1 Internal Gosub(app-missedcall-hangup, 479883994@from,1) start
  == Spawn extension (dial-with-exten, 306, 1) exited non-zero on 'Local/479883994@from-internal-000000b3;1'
    -- Local/479883994@from-internal-000000b3;1 Internal Gosub(crm-hangup,s,1) complete GOSUB_RETVAL=
    -- Channel Local/479883994@from-internal-000000b3;2 left 'simple_bridge' basic-bridge <c6256960-02fb-44af-ac4d-8d0900ba34de>
  == Spawn extension (trunk-dial-with-exten, 479883994, 1) exited non-zero on 'Local/479883994@from-internal-000000b3;2'
    -- Channel SIP/VI_OUT-0000008a left 'simple_bridge' basic-bridge <c6256960-02fb-44af-ac4d-8d0900ba34de>
    -- SIP/VI_OUT-0000008a Internal Gosub(app-missedcall-hangup,VI_OUT,1) start
      == Spawn extension (from-internal, 306, 1) exited non-zero on 'Local/RG-306*-479883994#@from-internal-000000b2;1'
    -- Local/RG-306*-479883994#@from-internal-000000b2;1 Internal Gosub(crm-hangup,s,1) complete GOSUB_RETVAL=
    -- No one is available to answer at this time (1:0/0/0)

**Comment by KSRyanMerritt** [comment]
> I've got a ticket open with Sangoma, but they're still bouncing around with requesting PCAPs and telling me my ring group is setup wrong. They're supposed to be escalating to dev team as well, there were some updates to core over the weekend pushed to public, so I will test again after hours removing my fix and letting dialparties.agi overwrite to see if its been fixed.

While we are waiting on this fix, would you be willing to share you work around with me so I can get mine working?

**Comment by ravnx** [comment]
KSRyanMerritt,
My fix is at the top in the bug submission, it was only for the number not
getting dialed due to the characters, { }. I didnt notice a confirm not
picking up, I hadnt had a chance to patch the system yet either. You might
turn on DTMF debugging and make sure its actually hearing the DTMF.

**Comment by kapilgupta01** [comment]
Hi @KSRyanMerritt are you still the same issue or facing some other issue?

**Comment by KSRyanMerritt** [comment]
We now have a new issue: As soon as you press 1 on the remote phone to confirm that you accept the call, the PBX hangs up on the remote caller and sends the originating caller to the invalid destination.

**Comment by jissphilip** [comment]
@KSRyanMerritt  Please test with ringgroups version 17.0.2.6

**Comment by KSRyanMerritt** [comment]
Thank you ringgroups version 17.0.2.6 seems to have fixed it thanks!
